// npm/lib/utils/find.js
/*
 * Meran - MERAN UNLP is a ILS (Integrated Library System) wich provides Catalog,
 * Circulation and User's Management. It's written in Perl, and uses Apache2
 * Web-Server, MySQL database and Sphinx 2 indexing.
 * Copyright (C) 2009-2013 Grupo de desarrollo de Meran CeSPI-UNLP
 *
 * This file is part of Meran.
 *
 * Meran is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Meran is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Meran.  If not, see <http://www.gnu.org/licenses/>.
 */


// walks a set of directories recursively, and returns
// the list of files that match the filter, if one is
// provided.

module.exports = find
var fs = require("graceful-fs")
  , asyncMap = require("slide").asyncMap
  , path = require("path")

function find (dir, filter, depth, cb) {
  if (typeof cb !== "function") cb = depth, depth = Infinity
  if (typeof cb !== "function") cb = filter, filter = null
  if (filter instanceof RegExp) filter = reFilter(filter)
  if (typeof filter === "string") filter = strFilter(filter)
  if (!Array.isArray(dir)) dir = [dir]
  if (!filter) filter = nullFilter
  asyncMap(dir, findDir(filter, depth), cb)
}
function findDir (filter, depth) { return function (dir, cb) {
  fs.lstat(dir, function (er, stats) {
    // don't include missing files, but don't abort either
    if (er) return cb()
    if (!stats.isDirectory()) return findFile(dir, filter, depth)("", cb)
    var found = []
    if (!filter || filter(dir, "dir")) found.push(dir+"/")
    if (depth <= 0) return cb(null, found)
    cb = (function (cb) { return function (er, f) {
      cb(er, found.concat(f))
    }})(cb)
    fs.readdir(dir, function (er, files) {
      if (er) return cb(er)
      asyncMap(files, findFile(dir, filter, depth - 1), cb)
    })
  })
}}
function findFile (dir, filter, depth) { return function (f, cb) {
  f = path.join(dir, f)
  fs.lstat(f, function (er, s) {
    // don't include missing files, but don't abort either
    if (er) return cb(null, [])
    if (s.isDirectory()) return find(f, filter, depth, cb)
    if (!filter || filter(f, "file")) cb(null, f)
    else cb(null, [])
  })
}}
function reFilter (re) { return function (f, type) {
  return nullFilter(f, type) && f.match(re)
}}
function strFilter (s) { return function (f, type) {
  return nullFilter(f, type) && f.indexOf(s) === 0
}}
function nullFilter (f, type) { return type === "file" && f }
