// node-gyp/lib/clean.js
/*
 * Meran - MERAN UNLP is a ILS (Integrated Library System) wich provides Catalog,
 * Circulation and User's Management. It's written in Perl, and uses Apache2
 * Web-Server, MySQL database and Sphinx 2 indexing.
 * Copyright (C) 2009-2013 Grupo de desarrollo de Meran CeSPI-UNLP
 *
 * This file is part of Meran.
 *
 * Meran is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Meran is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Meran.  If not, see <http://www.gnu.org/licenses/>.
 */


module.exports = exports = clean

exports.usage = 'Removes any generated build files and the "out" dir'

/**
 * Module dependencies.
 */

var rm = require('rimraf')
  , asyncEmit = require('./util/asyncEmit')
  , createHook = require('./util/hook')


function clean (gyp, argv, callback) {

  // Remove the 'build' dir
  var buildDir = 'build'
    , emitter

  createHook('gyp-clean.js', function (err, _e) {
    if (err) return callback(err)
    emitter = _e
    asyncEmit(emitter, 'before', function (err) {
      if (err) return callback(err)
      doClean()
    })
  })

  function doClean () {
    gyp.verbose('removing "build" directory')
    rm(buildDir, after)
  }

  function after () {
    asyncEmit(emitter, 'after', function (err) {
      if (err) return callback(err)
      callback()
    })
  }

}
